### PR TITLE
More testing of the Java crypto implementation

### PR DESCRIPTION
**Pull request by QuintinWillison**
See ably/ably-cocoa#921 for source of the data in 4f2dd0d.

Tests fail and, while that is uexpected, it is the current state of this branch - where it could either be a problem with my testing method or it could be a bug in the crypto code. Specifically:

- io.ably.lib.test.realtime.`RealtimeCryptoMessageTest`:
  - `encrypt_message_128`
  - `encrypt_message_256`
- io.ably.lib.test.realtime.`RealtimeCryptoTest`:
  - `decodeAppleLibrarySequences`

#531

### CODE REVIEW

**Comment by paddybyers** [comment]
See https://github.com/ably/ably-java/tree/531-encryption-investigations-paddy. All test cases after the first fail, because the act of using the `CipherParams` in one test mutates the `iv` for subsequent tests. That shouldn't happen of course, but that's the reason for the failures. In practice clients won't specify an `iv`, which is why we've not seen that issue in the wild.

File: lib/src/main/java/io/ably/lib/util/Crypto.java [hunk 1]
```diff
@@ -111,12 +129,10 @@ public static CipherParams getDefaultParams(String base64Key) {
 	}
 
 	/**
-	 * Obtain default CipherParams using key and algorithm from other CipherParams object
-	 * @param params
-	 * @return
+	 * Package scoped method for unit testing purposes.
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/util/Crypto.java]: This is a public API: https://docs.ably.io/client-lib-development-guide/features/#RSE1

File: lib/src/test/java/io/ably/lib/test/realtime/RealtimeCryptoMessageTest.java [hunk 2]
```diff
@@ -37,6 +38,12 @@ public void encrypt_message_256() throws IOException, NoSuchAlgorithmException,
 		testEncrypt(testData, "cipher+aes-256-cbc");
 	}
 
+	@Test
+	public void encryptCocoa() throws IOException, NoSuchAlgorithmException, CloneNotSupportedException, AblyException {
```
Comment (paddybyers) [inline lib/src/test/java/io/ably/lib/test/realtime/RealtimeCryptoMessageTest.java]: Why are these identified as Cocoa fixtures? Aren't they just common fixtures?

File: lib/src/test/java/io/ably/lib/util/CryptoTest.java [hunk 3]
```diff
@@ -63,4 +69,108 @@ public void cipher_params() throws AblyException, NoSuchAlgorithmException {
...
+		writeData(writer, encoded, null);
+
+		writer.name("encrypted");
+		writeData(writer, encoded, "cipher+aes-256-cbc");
+
+		writer.endObject();
+	}
+
+	private static final String BASE64 = "base64";
+	private void writeData(final JsonWriter writer, final byte[] data, final String encoding) throws IOException {
```
Comment (paddybyers) [inline lib/src/test/java/io/ably/lib/util/CryptoTest.java]: Doesn't there need to be something in here that writes the msgpack-formatted message as well?

Comment (QuintinWillison) [reply lib/src/test/java/io/ably/lib/util/CryptoTest.java]: Nothing that I've touched yet that has to read these has needed that property.

Comment (QuintinWillison) [reply lib/src/test/java/io/ably/lib/util/CryptoTest.java]: Nothing that I've touched, so far, has needed that property.

File: lib/src/test/java/io/ably/lib/test/realtime/RealtimeCryptoMessageTest.java (hunk 2 above)
Comment (QuintinWillison) [reply lib/src/test/java/io/ably/lib/test/realtime/RealtimeCryptoMessageTest.java]: Now gone. Any chance you can revisit this PR, please, @paddybyers ? I'm seeking an approval prior to merge. Thanks.

File: lib/src/test/java/io/ably/lib/util/CryptoMessageTest.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+		final CipherParams params = Crypto.getParams(algorithm, fixtureSet.key, fixtureSet.iv);
+		final ChannelOptions options = new ChannelOptions() {{encrypted = true; cipherParams = params;}};
+
+		for(final CryptoTestItem item : testData.items) {
+			final Message plain = item.encoded;
+			final Message encrypted = item.encrypted;
+			assertThat(encrypted.encoding, endsWith(fixtureSet.cipherName + "/base64"));
+
+			// if necessary, remove base64 encoding from plain-'text' message
```
Comment (paddybyers) [inline lib/src/test/java/io/ably/lib/util/CryptoMessageTest.java]: Comment syntax is inconsistent with other changes in this commit. In this repo we've been using `/* .... */`-style comments generally.

Comment (QuintinWillison) [reply lib/src/test/java/io/ably/lib/util/CryptoMessageTest.java]: Thanks. I'm going to leave this as is - it'll get picked up, if necessary, when we get around to conforming code style across the codebase. For now I am happy that, albeit in a minority, my commenting style isn't entirely non-standard even within this codebase:

```
ably-java % git grep '/\* ' | wc -l
    2332
ably-java % git grep '// ' | wc -l
     191
```

Comment (QuintinWillison) [reply lib/src/test/java/io/ably/lib/util/CryptoMessageTest.java]: Issue created: #537

**Comment by QuintinWillison** [comment]
I've squashed a subset of the commits and then merged manually.

**Pull request closed without merging**